### DISCUSSION TITLE
App doesn't respond after loading a complex route

### DISCUSSION THREAD

**Discussion by WolfgangZ42** [devemux86/cruiser#638] [Q&A] [answered]
When I load a relatively complex route, it often happens that Cruiser no longer responds. 
The route is loaded - the info box pops up, but a few seconds later the corresponding display appears. 

<img width="200" src="https://github.com/user-attachments/assets/13568080-3e4d-4301-ac4c-e08c59102aac"/> . . . <img width="200" src="https://github.com/user-attachments/assets/9d625834-b641-4977-9f41-0dc61efbfd78"/>

After “Close app” and calling up Cruiser again, the route is loaded correctly. Here is an example route and my profile: [Example.zip](https://github.com/user-attachments/files/17268231/Example.zip) - I hope you can reproduce it.

**Comment by devemux86** [comment]
@WolfgangZ42 Thanks for the report.

I cannot reproduce it using your profile and route example.

How do you import the route?
With the menu "Routing | Import"?

**Reply by devemux86** [reply]
@WolfgangZ42 

Your screenshots do not show the loaded route, only overlays?

How many favorites and routes / tracks do you show on the map?
Perhaps they are too many?

**Reply by bios4** [reply]
@devemux86

I have to chime in also in this thread - I had the exact same error as @WolfgangZ42 a few days ago!

Because this was - for me at this time - a one-time occurence and the route was properly loaded after a restart of Cruiser I did not think that this was a general issue in the app, but now as there is another user with the very same observations there might be really something wrong...

> How do you import the route?

In my case the route (a *.kurviger file) was loaded/imported by long-pressing the navigation button > Import.
There were no other routes/tracks displayed/loaded at that time.

**Reply by WolfgangZ42** [reply]
Route is loaded via **/Routing/Import**
The loaded route only shows up after restarting Cruiser.
And Overlays are (Favorites/Tracks):

<img width="200" src="https://github.com/user-attachments/assets/8aaba7a1-3dd7-4d27-b810-c0fa96a869b5"/> . . . <img width="200" src="https://github.com/user-attachments/assets/79beaad2-82d0-4c71-8e35-93ab13cf553e"/>

**Reply by WolfgangZ42** [reply]
I would like to add that I don't normally use so many waypoints - but due to the preference for dirt roads and parts of the Portugal TET, they are necessary.

**Comment by bios4** [comment]
I had this issue with the frozen Cruiser app again after loading today's "Cruiser 5 test"-route!!!

- Opened Cruiser (5.0.0-b23 that is)
- Imported my *.kurviger route (+ overlay) for today via long-press navigation button > Import
- Cruiser froze immediately with the message shown below
- Force-closed the app, deleted its cache in the Android apps menu
- Opened Cruiser again
- Imported the same *.kurviger route (+ overlay) again (this time via Routing > Import > Kurviger option "Overlay + Routing"), with the same result - Cruiser immediately crashed/stopped responding
- Closed the app but without deleting the cache
- Opened Cruiser again - the route was displayed but without the previously also imported overlay
- Started navigation (without the overlay)
- Cruiser behaved weird for a few hundred meters (not properly loading the map, not refreshing the location), but suddenly recovered and everything worked as expected before I could stop again, so I continued driving

<img src="https://github.com/user-attachments/assets/ea8c5099-3e2c-4969-b4e8-1aacef18edb5" width="33%">

Should I send you the route in question and the profile I used @devemux86?
I don't want to share the route publicly...

**Reply by devemux86** [reply]
> Opened Cruiser again - the route was displayed but without the previously also imported overlay

Overlays are saved if they are imported in "My collection".

**Reply by devemux86** [reply]
@bios4 Please send them.

It is important to reproduce a report, otherwise it is difficult to check what happens.

**Reply by bios4** [reply]
> Overlays are saved if they are imported in "My collection".

I know; I just wanted to explain why I imported the route "from scratch" again (because I wanted to have the overlay also on the map)...

> Please send them.

You've got mail!

**Reply by devemux86** [reply]
@bios4 Thanks for the files.

Unfortunately again I cannot reproduce the report.

How many favorites and routes / tracks do you show on the map?

**Reply by bios4** [reply]
> How many favorites and routes / tracks do you show on the map?

- 6 tracks for "95 dB closures" in Austria, total 143 km
- 1 very tiny "no go" area in my hometown indicating a long-term construction site
- 12 favorites for the "95 db closures"
- 54 total other favorites separated in 4 categories/groups

**Comment by devemux86** [comment]
Questions to both @WolfgangZ42 and @bios4:

- What happens when you load a more simple route?
Does it really depend on the complexity of the route?

- What happens when you have active only 1 map of the route area?
(as you both seem to have loaded many offline maps)

- What happens if hillshading is disabled?
(and reload the map theme)

**Reply by bios4** [reply]
> What happens when you load a more simple route?
Does it really depend on the complexity of the route?

I'm sorry but I cannot answer this. 😞

- Loaded today's route causing the reported issue once more - Cruiser crashed
- Restarted Cruiser, deleted the initial route and loaded a shorter test route (approx. 6 km) - no issue...
- Loaded the initial route again - no issues anymore...
- Closed Cruiser once more, deleted the cache
- Loaded the initial route a third time - no issues...

Weird.

I did not check the other points because the problem isn't reproducible anymore (at least right now).

**Reply by bios4** [reply]
Should I update to 5.0.0-b25 or stay at 5.0.0-b23 for further investgations @devemux86?

**Reply by devemux86** [reply]
@bios4 

App not responding is completely different from a crash. What do you see?

> Should I update to 5.0.0-b25

Definitely, we can only test and fix with the latest code.

**Reply by bios4** [reply]
> Not responding is completely different from a crash. What do you see?

After loading of the troublesome route to verify your points the behaviour was exactly the same as in my error report - Cruiser froze and showed "Not responding / close app" after a short period of time as it did during my test drive (shown in the screenshot above).

But this did not occur anymore after I imported the (shorter) test route.
Also when loading the erroneus route again.

> > Should I update to 5.0.0-b25

> Definitely, we can only test and fix with the latest code.

OK, will do tomorrow...

**Reply by WolfgangZ42** [reply]
Only with Germany-Map and without hillshading the complex routes failed - simple one works. 
I don't find any coherence ... 
Because everything is okay after restarting Cruiser, I can live with it at moment.

**Reply by devemux86** [reply]
@bios4 

So it is "not responding".
A crash is different and closes an app immediately without "not responding" message.

We need to use the proper terms to understand what really happens. :slightly_smiling_face:

**Reply by devemux86** [reply]
@WolfgangZ42 

Your example is from Portugal. Do you have a route example from Germany?

Can you reproduce it each time?

**Reply by WolfgangZ42** [reply]
I found a German route: 
[0699_München SW.kurviger.txt](https://github.com/user-attachments/files/17270174/0699_Munchen.SW.kurviger.txt)
And yes it's happening each import now.

**Comment by devemux86** [comment]
@WolfgangZ42 @bios4 

The problem is probably with the automatic map zoom / scroll to show the whole imported route.

I will think how we can solve this.

---

If you keep the map at same area / zoom after import a complex route.
Clear the route.
Exit the app and remove it from recent apps.
Import again the complex route, does it work?

But if at the 1st step, move / zoom to a completely different map area.
And follow the next steps, then you could probably reproduce it?
(as the map will zoom / move to the imported route area)

**Reply by WolfgangZ42** [reply]
BINGO - you got it!

**Reply by bios4** [reply]
Nope, not on my device... 😞

- Imported the troublesome route again: Cruiser froze/showed "Not responding" / "Close" again
- Restarted Cruiser
- Cleared the route, moved visible map area to a totally different location
- Exited the app and cleared it from recent apps
- Started Cruiser again
- Imported the troublesome route a second time: import & zoom to show the whole root worked, no issue

Still on b23, by the way (because I'm not at home and can't update right now).

**Reply by devemux86** [reply]
@bios4 

> no issue

You see that the opposite works, which is fine:
Be at a different map area, import route, map moves to show the route.

We are looking at what happens when the map is already in the route area
and we import the route again, does it happen?
(as the map does not need to zoom / move)

**Reply by WolfgangZ42** [reply]
> BINGO - you got it!

I was able to reproduce it, believe me - now it works even when the map area is completely different ... strange ...

**Reply by devemux86** [reply]
@WolfgangZ42 

Such events do not always happen, so it is difficult to catch and understand them.

I will look at this case and we will see if it solves it on your devices.
(as here until now it works every time)

**Reply by bios4** [reply]
> We are looking at what happens when the map is already in the route area
and we import the route again, does it happen?

Nope, I can't reproduce it again with either method.
It "works" right now in every condition...

These were my last attempts with b23 - I will update to b25 tomorrow and then we'll see further.

**Comment by devemux86** [comment]
@WolfgangZ42 @bios4 

Please try the latest Beta 26.

**Reply by bios4** [reply]
- Installed it right away
- Loaded my troublesome route from yesterday, map area NOT at the location of the route
- Cruiser froze and showed the known error message

😢

<img src="https://github.com/user-attachments/assets/90117b7c-76c8-45d3-acb1-d227b91151a2" width="33%">

**Reply by bios4** [reply]
This happened now TWICE, also after restart Cruiser > delete route > close Cruiser > remove it from recent apps > restart Cruiser > load troublesome route.

**Reply by bios4** [reply]
And ONCE AGAIN after the same procedure ^^, but now with the visible map area at the location of the route.

The app freezes now reproducibly right after the modal with the route data (length, duration etc.) is shown...

**Reply by devemux86** [reply]
Unfortunately I cannot reproduce it, so it is not possible to check what happens.

I will prepare another update disabling the zoom to the route bounds and see.

**Reply by devemux86** [reply]
Also we already have many months with beta testing.

It is strange that this happens only now.

**Reply by bios4** [reply]
As I wrote above this happened to me ONCE a few beta versions earlier.
And then no more until my ride on Sunday...

**Comment by devemux86** [comment]
@WolfgangZ42 @bios4 

Please try the latest Beta 27.

**Reply by bios4** [reply]
Attempt 1:
- Installed b27
- Loaded my troublesome route from yesterday (map area at the location of the route)
- No issues, route data modal (length, duration...) is shown and Cruiser allows starting the navigation

Attempt 2:
- Deleted the route
- Moved map area outside the location of the route
- Closed Cruiser
- Removed it from recent apps
- Restarted Cruiser
- Loaded my troublesome route from yesterday (map area NOT at the location of the route)
- No issues, route data modal (length, duration...) is shown and Cruiser allows starting the navigation

While it is "strange" that the route is not shown on the screen after import I could live with it if this really fixes the issue.

Please note: I can't "really" navigate because I'm at work, so I can't elaborate about any possible weird behaviour or freezes after starting a ride (like seen yesterday)...

**Reply by devemux86** [reply]
@bios4 

> While it is "strange" that the route is not shown on the screen after import I could live with it if this really fixes the issue.

The route is rendered on the map, but the map does not automatically moves on the route bounds after import.
This was currently disabled, until we understand what happens, because I cannot reproduce it yet. :slightly_frowning_face: 

Note: there is always the "Show whole route" action when long press the map.

**Reply by WolfgangZ42** [reply]
Same here - no issues 
And **"tap on any map location"/More/Show whole route** - here we are!

**Reply by bios4** [reply]
> Note: there is always the "Show whole route" action when long press the map.

I just want to understand: what makes this ^^ sequence of commands different from automatically displaying the entire route after import? Because when run manually, it doesn't seem to cause any problems - at least not currently, without further testing “on the road”...

But again: if this solves the issue, I'm fine. 😉

**Reply by devemux86** [reply]
@bios4 This is what we need to find, because here it works as expected, so it should work everywhere.

**Comment by devemux86** [comment]
Import route displays "Show whole route" prompt after import in **Cruiser 5.0.1-b8**:
- https://github.com/devemux86/cruiser/discussions/641#discussioncomment-10982507

**Reply by devemux86** [reply]
@WolfgangZ42 Can you try the import with **5.0.1-b8**?

**Reply by WolfgangZ42** [reply]
Sure - but I'm right now on (or better off) road in Portugal. So in 10 days I'll able to test it. Sorry for that.

**Reply by devemux86** [reply]
@WolfgangZ42 No problem!

**Reply by devemux86** [reply]
@bios4 Can you also try the import with **5.0.1-b8**?

**Reply by bios4** [reply]
Tested it right now with 5.0.1-**b9**...

Used three different routes of my collection from different folders and with different lengths, where two of them caused troubles in the past. Everything worked ok for both settings (either when confirming "Show whole route" or when declining the prompt).

BUT: when a very short route is loaded, the modal popup showing route length, duration etc. is displayed even BEFORE I select the possible options. I assumed the route to be loaded only AFTER my desired option is confirmed, but not before.

Is this intended behaviour @devemux86?
Or would it be better to wait until the user selects the desired option and load the route only then?

**Reply by devemux86** [reply]
The temporary message with the route details appears when the route is loaded / imported.
At the same time the modal question appears to show the whole route or not.

The route is loaded anyway.
The question is whether to zoom the map to the route bounds or not.

**Reply by MotoUKRider** [reply]
@devemux86 and @bios4 How about incorporating the route info into the modal dialog?
(assuming that it remains in place)

My thinking being that, whether you want to see the route or not - at that point in the workflow, you only have 1 place to be looking at, rather than a fleeting popup at the bottom of the screen.

**Reply by devemux86** [reply]
@MotoUKRider I will check it.

**Reply by bios4** [reply]
> How about incorporating the route info into the modal dialog?

I like that suggestion!
Maybe Emux can find a suitable solution...

**Reply by WolfgangZ42** [reply]
Now I tested several files - everything works, as you already know. 
Thanks for solving it!

**Comment by DolomitesTom** [comment]
B9: On a 200 km tour four times "App not responding". However, this time it didn't crash and I was able to continue riding. Could it be that the average speed is reset after each app restart? I don't think that's ideal. It would be better to reset it manually as before.

**Comment by DolomitesTom** [comment]
b10: still 3 times on a 100km Tour, the Message appears :
ePA doesn't react
The message disappears after a while.

**Reply by devemux86** [reply]
Attach here your profile and the route.

**Comment by DolomitesTom** [comment]
This is a very personal route from which you can read my home town, so I don't want to do that. I drive it 2-3 times a week. But it also happens on other routes. It could happen if you deliberately deviate from the planned route and a recalculation takes place and you don't adjust the route immediately by skipping it. Then a multiple recalculation takes place. Could this take too long and result in timeouts? The app no longer crashes, but the message disappears again after approx. 10 to 15 seconds. So it is not that critical. It's just a slight blur. And I still drive more often with V4. But it only happens with V5.0.1 bXX, actually b10. It is also possible that it happens when routes cross each other, just an assumption. 

Next Sunday, for example, I'm driving a tour that I could send if it happens again. It is a non-intersecting route. I will use V4 and V5 in parallel to see what happens.

By the way, could the reset of the average speed be optional on every closure of the app? I often close the app in pauses, but want to keep the average speed of the whole trip. Or was it a wish of a user to reset this every time automatically?

**Reply by devemux86** [reply]
> could the reset of the average speed be optional on every closure of the app?

Reset of average speed can only happen when users press the average speed panel.
Otherwise the average speed is saved and remains the same when restart the app.

**Reply by DolomitesTom** [reply]
Hmm, right, don't know why I thought so, wrong impression 🤔.

**Comment by devemux86** [comment]
**[Cruiser 5.0.2-b1](https://github.com/devemux86/cruiser/discussions/641#discussioncomment-11134244)** contains changes trying to solve the not responding report.
The zoom after import is again automatic (without question) to check if things have been fixed.

@WolfgangZ42 @bios4 Can you test this version?

---

(the previous version is available in [Releases](https://github.com/devemux86/cruiser/releases), if you want to restore it)

**Reply by WolfgangZ42** [reply]
Did try several complex routes - no problem! Thanks for fixing it!

**Reply by devemux86** [reply]
If this works, then it should also fix the not responding reports for navigation.

**Reply by bios4** [reply]
When importing/loading *.kurviger routes, which caused troubles earlier, from the file system the issue seems to be resolved! 👍
When importing stored Cruiser GPS-recordings as *.gpx with trkseg, everything seems to work, too.

But when loading the very same GPS-recordings from "My collection" (swipe right), only the first trkseg is loaded as a route.
See attached screenshots:

1. Route in "My collection"
2. Route when "Show on map" is activated
3. Route when loaded/imported via swipe-right gesture
4. Route in total overview

<img src="https://github.com/user-attachments/assets/42b51dfe-2c1a-4584-bbfe-5dd0c23017a3" width="50%">
<img src="https://github.com/user-attachments/assets/2da5d0e8-5ab1-4374-a8fb-cb92f19bc6a1" width="50%">
<img src="https://github.com/user-attachments/assets/1851a540-11e0-4ec8-bb40-afaa7f810012" width="50%">
<img src="https://github.com/user-attachments/assets/aef7b6c5-885c-4a7f-8fe5-c30e658a834c" width="50%">

Is this an intentional behavior?
Is this a user error?
Or is it a bug?

**Reply by bios4** [reply]
BTW this ^^ report applies to 5.0.2-b2 @devemux86!

**Reply by devemux86** [reply]
@bios4 Can you post the gpx file?

It is not possible to test without all available data. :slightly_smiling_face:

**Reply by bios4** [reply]
There you go...

[20240519_152640.gpx.txt](https://github.com/user-attachments/files/17617337/20240519_152640.gpx.txt)

**Reply by devemux86** [reply]
@bios4 Thanks for the report! :+1: 

This should be fixed in **5.0.2-b4**.

**Reply by bios4** [reply]
> This should be fixed in 5.0.2-b4.

Seems to work properly now (tested with at least two routes where the issue occured previously).
Thanks for the quick fix! 👍

**Comment by devemux86** [comment, marked as answer]
- Fixed in https://github.com/devemux86/cruiser/discussions/641
